// vue-virtual-scroller@2.0.0-beta.10/dist/vue-virtual-scroller.js
import { reactive as qe, toValue as u, provide as ze, computed as ie, watch as L, onActivated as je, onDeactivated as tt, onUnmounted as it, nextTick as ce, ref as Q, onMounted as Ie, onBeforeUnmount as Te, markRaw as nt, shallowReactive as ot, defineComponent as fe, openBlock as ee, createBlock as ae, resolveDynamicComponent as Re, withCtx as ne, renderSlot as te, createElementBlock as ue, withDirectives as rt, unref as K, normalizeClass as Le, createCommentVNode as ge, normalizeStyle as st, Fragment as lt, renderList as ut, mergeProps as _e, toHandlers as ct, createVNode as at, createSlots as ft, normalizeProps as dt, guardReactiveProps as mt, inject as he, getCurrentInstance as vt, onBeforeUpdate as pt } from "vue";
import yt from "mitt";
function zt(e, n, r, t) {
  let i = 0, c = {};
  const w = yt();
  let S = !1, y;
  const g = qe({
    active: !0,
    sizes: {},
    keyField: u(e).keyField,
    simpleArray: !1
  });
  typeof ResizeObserver < "u" && (y = new ResizeObserver((s) => {
    requestAnimationFrame(() => {
      if (Array.isArray(s)) {
        for (const d of s)
          if (d.target && d.target.$_vs_onResize) {
            let z, R;
            if (d.borderBoxSize) {
              const h = d.borderBoxSize[0];
              z = h.inlineSize, R = h.blockSize;
            } else
              z = d.contentRect.width, R = d.contentRect.height;
            d.target.$_vs_onResize(d.target.$_vs_id, z, R);
          }
      }
    });
  })), ze("vscrollData", g), ze("vscrollParent", {
    get $_undefinedSizes() {
      return i;
    },
    set $_undefinedSizes(s) {
      i = s;
    },
    get $_undefinedMap() {
      return c;
    },
    set $_undefinedMap(s) {
      c = s;
    },
    $_events: w,
    direction: ie(() => u(e).direction)
  }), ze("vscrollResizeObserver", y);
  const b = ie(() => {
    const s = u(e);
    return s.items.length > 0 && typeof s.items[0] != "object";
  }), v = ie(() => {
    const s = [], d = u(e), { items: z, keyField: R } = d, h = b.value, U = g.sizes, a = z.length;
    for (let _ = 0; _ < a; _++) {
      const P = z[_], I = h ? _ : P[R];
      let N = U[I];
      typeof N > "u" && !c[I] && (N = 0), s.push({
        item: P,
        id: I,
        size: N
      });
    }
    return s;
  });
  function E() {
    var s;
    u(n) && x(), (s = t == null ? void 0 : t.onResize) == null || s.call(t);
  }
  function C() {
    var s;
    w.emit("vscroll:update", { force: !1 }), (s = t == null ? void 0 : t.onVisible) == null || s.call(t);
  }
  function x(s = !1) {
    (s || b.value) && (g.sizes = {}), w.emit("vscroll:update", { force: !0 });
  }
  function $(s) {
    const d = u(n);
    d && d.scrollToItem(s);
  }
  function H(s, d) {
    const z = u(e), R = b.value ? d ?? z.items.indexOf(s) : s[z.keyField];
    return g.sizes[R] || 0;
  }
  function D() {
    const s = u(r);
    s && (S || (S = !0, ce(() => {
      s.scrollTop = s.scrollHeight + 5e3;
      const d = () => {
        s.scrollTop = s.scrollHeight + 5e3, requestAnimationFrame(() => {
          s.scrollTop = s.scrollHeight + 5e3, i === 0 ? S = !1 : requestAnimationFrame(d);
        });
      };
      requestAnimationFrame(d);
    })));
  }
  return L(() => u(e).items, () => {
    x();
  }), L(b, (s) => {
    g.simpleArray = s;
  }, { immediate: !0 }), L(() => u(e).direction, () => {
    x(!0);
  }), L(v, (s, d) => {
    const z = u(r);
    if (!z)
      return;
    const R = z.scrollTop, h = u(e);
    let U = 0, a = 0;
    const _ = Math.min(s.length, d.length);
    for (let I = 0; I < _ && !(U >= R); I++)
      U += d[I].size || h.minItemSize, a += s[I].size || h.minItemSize;
    const P = a - U;
    P !== 0 && (z.scrollTop += P);
  }), je(() => {
    g.active = !0;
  }), tt(() => {
    g.active = !1;
  }), it(() => {
    w.all.clear();
  }), {
    vscrollData: g,
    itemsWithSize: v,
    simpleArray: b,
    resizeObserver: y,
    forceUpdate: x,
    scrollToItem: $,
    getItemSize: H,
    scrollToBottom: D,
    onScrollerResize: E,
    onScrollerVisible: C
  };
}
const Ye = {
  itemsLimit: 1e3
}, gt = /auto|scroll/;
function Ge(e, n) {
  return e.parentNode === null ? n : Ge(e.parentNode, [...n, e]);
}
function Se(e, n) {
  return getComputedStyle(e, null).getPropertyValue(n);
}
function ht(e) {
  return Se(e, "overflow") + Se(e, "overflow-y") + Se(e, "overflow-x");
}
function St(e) {
  return gt.test(ht(e));
}
function He(e) {
  if (!(e instanceof HTMLElement || e instanceof SVGElement))
    return;
  const n = Ge(e.parentNode, []);
  for (let r = 0; r < n.length; r += 1)
    if (n[r] instanceof Element && St(n[r]))
      return n[r];
  return document.scrollingElement || document.documentElement;
}
let we = !1;
function _t() {
  return we;
}
if (typeof window < "u") {
  we = !1;
  try {
    const e = Object.defineProperty({}, "passive", {
      get() {
        we = !0;
      }
    });
    window.addEventListener("test", null, e);
  } catch {
  }
}
let wt = 0;
function $t(e, n, r, t, i) {
  const c = Q([]), w = Q(0), S = Q(!1);
  let y = 0, g = 0;
  const b = /* @__PURE__ */ new Map(), v = /* @__PURE__ */ new Map();
  let E = !1, C = 0, x = !1, $ = null, H = null, D = null, s = 0, d = null;
  const z = ie(() => {
    const o = u(e);
    return o.items.length > 0 && typeof o.items[0] != "object";
  }), R = ie(() => {
    const o = u(e);
    if (o.itemSize === null) {
      const l = {
        [-1]: { accumulator: 0 }
      }, f = o.items, m = o.sizeField, V = o.minItemSize;
      let j = 1e4, F = 0, W;
      for (let Z = 0, re = f.length; Z < re; Z++)
        W = f[Z][m] || V, W < j && (j = W), F += W, l[Z] = { accumulator: F, size: W };
      return s = j, l;
    }
    return [];
  });
  function h(o) {
    let l = v.get(o);
    return l || (l = [], v.set(o, l)), l;
  }
  function U(o, l, f, m, V) {
    const j = nt({
      id: wt++,
      index: l,
      used: !0,
      key: m,
      type: V
    }), F = ot({
      item: f,
      position: 0,
      offset: 0,
      nr: j
    });
    return o.push(F), F;
  }
  function a(o) {
    const l = h(o);
    if (l && l.length) {
      const f = l.pop();
      return f.nr.used = !0, f;
    }
  }
  function _(o) {
    const l = o.nr.type;
    h(l).push(o), o.nr.used = !1, o.position = -9999, b.delete(o.nr.key);
  }
  function P() {
    b.clear(), v.clear();
    for (let o = 0, l = c.value.length; o < l; o++)
      _(c.value[o]);
  }
  function I() {
    var o;
    (o = i == null ? void 0 : i.onResize) == null || o.call(i), S.value && q(!1);
  }
  function N() {
    const o = u(e);
    if (!E) {
      if (E = !0, $)
        return;
      const l = () => requestAnimationFrame(() => {
        E = !1;
        const { continuous: f } = q(!1, !0);
        f || (H && clearTimeout(H), H = setTimeout(N, o.updateInterval + 100));
      });
      l(), o.updateInterval && ($ = setTimeout(() => {
        $ = null, E && l();
      }, o.updateInterval));
    }
  }
  function Xe(o, l) {
    var f, m;
    S.value && (o || l.boundingClientRect.width !== 0 || l.boundingClientRect.height !== 0 ? ((f = i == null ? void 0 : i.onVisible) == null || f.call(i), requestAnimationFrame(() => {
      q(!1);
    })) : (m = i == null ? void 0 : i.onHidden) == null || m.call(i));
  }
  function Je() {
    const o = He(u(n));
    return window.document && (o === window.document.documentElement || o === window.document.body) ? window : o || window;
  }
  function be() {
    const o = u(n), l = u(e), f = l.direction === "vertical";
    let m;
    if (l.pageMode) {
      const V = o.getBoundingClientRect(), j = f ? V.height : V.width;
      let F = -(f ? V.top : V.left), W = f ? window.innerHeight : window.innerWidth;
      F < 0 && (W += F, F = 0), F + W > j && (W = j - F), m = {
        start: F,
        end: F + W
      };
    } else f ? m = {
      start: o.scrollTop,
      end: o.scrollTop + o.clientHeight
    } : m = {
      start: o.scrollLeft,
      end: o.scrollLeft + o.clientWidth
    };
    return m;
  }
  function Me() {
    u(e).pageMode ? Qe() : xe();
  }
  function Qe() {
    d = Je(), d.addEventListener("scroll", N, _t() ? { passive: !0 } : !1), d.addEventListener("resize", I);
  }
  function xe() {
    d && (d.removeEventListener("scroll", N), d.removeEventListener("resize", I), d = null);
  }
  function q(o, l = !1) {
    var Ce;
    const f = u(e), m = f.itemSize, V = f.gridItems || 1, j = f.itemSecondarySize || m, F = s, W = f.typeField, Z = z.value ? null : f.keyField, re = f.items, B = re.length, X = R.value, Ee = b, ye = c.value;
    let O, M, se, J, k;
    if (!B)
      O = M = J = k = se = 0;
    else if (x)
      O = J = 0, M = k = Math.min(f.prerender, re.length), se = 0;
    else {
      const p = be();
      if (l) {
        let Y = p.start - C;
        if (Y < 0 && (Y = -Y), m === null && Y < F || m !== null && Y < m)
          return {
            continuous: !0
          };
      }
      C = p.start;
      const le = f.buffer;
      p.start -= le, p.end += le;
      let A = 0;
      const me = u(r);
      me && (A = me.scrollHeight, p.start -= A);
      const Pe = u(t);
      if (Pe) {
        const Y = Pe.scrollHeight;
        p.end += Y;
      }
      if (m === null) {
        let Y, Be = 0, Oe = B - 1, G = ~~(B / 2), Ae;
        do
          Ae = G, Y = X[G].accumulator, Y < p.start ? Be = G : G < B - 1 && X[G + 1].accumulator > p.start && (Oe = G), G = ~~((Be + Oe) / 2);
        while (G !== Ae);
        for (G < 0 && (G = 0), O = G, se = X[B - 1].accumulator, M = G; M < B && X[M].accumulator < p.end; M++) ;
        for (M === -1 ? M = re.length - 1 : (M++, M > B && (M = B)), J = O; J < B && A + X[J].accumulator < p.start; J++) ;
        for (k = J; k < B && A + X[k].accumulator < p.end; k++) ;
      } else {
        O = ~~(p.start / m * V);
        const Y = O % V;
        O -= Y, M = Math.ceil(p.end / m * V), J = Math.max(0, Math.floor((p.start - A) / m * V)), k = Math.floor((p.end - A) / m * V), O < 0 && (O = 0), M > B && (M = B), J < 0 && (J = 0), k > B && (k = B), se = Math.ceil(B / V) * m;
      }
    }
    M - O > Ye.itemsLimit && Ze(), w.value = se;
    let T;
    const Fe = O <= g && M >= y;
    if (!Fe || o)
      P();
    else
      for (let p = 0, le = ye.length; p < le; p++)
        if (T = ye[p], T.nr.used) {
          const A = T.nr.index >= O && T.nr.index < M, me = m || X[p] && X[p].size;
          (!A || !me) && _(T);
        }
    let oe, de;
    for (let p = O; p < M; p++) {
      if (!(m || X[p] && X[p].size))
        continue;
      oe = re[p];
      const A = Z ? oe[Z] : p;
      if (A == null)
        throw new Error(`Key is ${A} on item (keyField is '${Z}')`);
      T = Ee.get(A), T ? (T.item !== oe && (T.item = oe), T.nr.used || console.warn(`Expected existing view's used flag to be true, got ${T.nr.used}`)) : (de = oe[W], T = a(de), T ? (T.item = oe, T.nr.index = p, T.nr.key = A, T.nr.type !== de && console.warn("Reused view's type does not match pool's type")) : T = U(ye, p, oe, A, de), Ee.set(A, T)), m === null ? (T.position = X[p - 1].accumulator, T.offset = 0) : (T.position = Math.floor(p / V) * m, T.offset = p % V * j);
    }
    return y = O, g = M, f.emitUpdate && ((Ce = i == null ? void 0 : i.onUpdate) == null || Ce.call(i, O, M, J, k)), D && clearTimeout(D), D = setTimeout(De, f.updateInterval + 300), {
      continuous: Fe
    };
  }
  function Ze() {
    throw setTimeout(() => {
      console.warn("It seems the scroller element isn't scrolling, so it tries to render all the items at once.", "Scroller:", u(n)), console.warn("Make sure the scroller has a fixed height (or width) and 'overflow-y' (or 'overflow-x') set to 'auto' so it can scroll correctly and only render the items visible in the scroll viewport.");
    }), new Error("Rendered items limit reached");
  }
  function ke() {
    const o = c.value.filter(({ nr: l }) => l.used);
    for (let l = 1; l < o.length; l++)
      if (o[l].nr.index !== o[l - 1].nr.index + 1)
        return !0;
    return !1;
  }
  function De() {
    c.value.sort((o, l) => o.nr.index - l.nr.index), ke() && (q(!1), D && clearTimeout(D));
  }
  function et(o) {
    const l = u(e);
    let f;
    const m = l.gridItems || 1;
    l.itemSize === null ? f = o > 0 ? R.value[o - 1].accumulator : 0 : f = Math.floor(o / m) * l.itemSize, ve(f);
  }
  function ve(o) {
    const l = u(e), f = u(n), m = l.direction === "vertical" ? { scroll: "scrollTop", start: "top" } : { scroll: "scrollLeft", start: "left" };
    if (l.pageMode) {
      const V = He(f), j = V.tagName === "HTML" ? 0 : V[m.scroll], F = V.getBoundingClientRect(), Z = f.getBoundingClientRect()[m.start] - F[m.start];
      V[m.scroll] = o + j + Z;
    } else
      f[m.scroll] = o;
  }
  const pe = u(e);
  return pe.prerender && (x = !0, q(!1)), pe.gridItems && !pe.itemSize && console.error("[vue-recycle-scroller] You must provide an itemSize when using gridItems"), Ie(() => {
    Me(), ce(() => {
      x = !1, q(!0), S.value = !0;
    });
  }), je(() => {
    const o = C;
    typeof o == "number" && ce(() => {
      ve(o);
    });
  }), Te(() => {
    xe();
  }), L(() => u(e).items, () => {
    q(!0);
  }), L(() => u(e).pageMode, () => {
    Me(), q(!1);
  }), L(R, () => {
    q(!1);
  }, { deep: !0 }), L(() => u(e).gridItems, () => {
    q(!0);
  }), L(() => u(e).itemSecondarySize, () => {
    q(!0);
  }), {
    pool: c,
    totalSize: w,
    ready: S,
    sizes: R,
    simpleArray: z,
    scrollToItem: et,
    scrollToPosition: ve,
    getScroll: be,
    updateVisibleItems: q,
    handleScroll: N,
    handleResize: I,
    handleVisibilityChange: Xe,
    sortViews: De
  };
}
const Ve = /* @__PURE__ */ new WeakMap();
function It(e) {
  return typeof e == "function" ? {
    callback: e,
    observer: null,
    intersection: void 0,
    visible: null
  } : {
    callback: e.callback,
    observer: null,
    intersection: e.intersection,
    visible: null
  };
}
function Ue(e, n) {
  Ke(e);
  const r = It(n.value);
  if (Ve.set(e, r), typeof IntersectionObserver > "u") {
    const t = e.getBoundingClientRect();
    r.visible = !0, r.callback(!0, {
      boundingClientRect: t
    });
    return;
  }
  r.observer = new IntersectionObserver((t) => {
    const i = t[0], c = !!(i != null && i.isIntersecting);
    r.visible !== null && r.visible === c || (r.visible = c, r.callback(c, i));
  }, r.intersection), r.observer.observe(e);
}
function Ke(e) {
  const n = Ve.get(e);
  n != null && n.observer && (n.observer.disconnect(), n.observer = null);
}
const Tt = {
  mounted(e, n) {
    Ue(e, n);
  },
  updated(e, n) {
    n.value !== n.oldValue && Ue(e, n);
  },
  unmounted(e) {
    Ke(e), Ve.delete(e);
  }
}, Rt = /* @__PURE__ */ fe({
  __name: "ItemView",
  props: {
    view: {},
    itemTag: {}
  },
  setup(e) {
    const n = e;
    return (r, t) => (ee(), ae(Re(n.itemTag), { class: "vue-recycle-scroller__item-view" }, {
      default: ne(() => [
        te(r.$slots, "default", {
          item: n.view.item,
          index: n.view.nr.index,
          active: n.view.nr.used
        })
      ]),
      _: 3
    }));
  }
}), Vt = /* @__PURE__ */ fe({
  __name: "ResizeObserver",
  emits: ["notify"],
  setup(e, { emit: n }) {
    const r = n, t = Q();
    let i = null, c = null;
    function w() {
      r("notify");
    }
    return Ie(() => {
      var y;
      const S = (y = t.value) == null ? void 0 : y.parentElement;
      if (S) {
        if (typeof ResizeObserver < "u") {
          i = new ResizeObserver(() => {
            w();
          }), i.observe(S);
          return;
        }
        c = () => w(), window.addEventListener("resize", c);
      }
    }), Te(() => {
      i && (i.disconnect(), i = null), c && (window.removeEventListener("resize", c), c = null);
    }), (S, y) => (ee(), ue("div", {
      ref_key: "el",
      ref: t,
      class: "vue-recycle-scroller__resize-observer",
      "aria-hidden": "true"
    }, null, 512));
  }
}), bt = (e, n) => {
  const r = e.__vccOpts || e;
  for (const [t, i] of n)
    r[t] = i;
  return r;
}, Mt = /* @__PURE__ */ bt(Vt, [["__scopeId", "data-v-08cc04ab"]]), $e = /* @__PURE__ */ fe({
  __name: "RecycleScroller",
  props: {
    items: {},
    keyField: { default: "id" },
    direction: { default: "vertical" },
    listTag: { default: "div" },
    itemTag: { default: "div" },
    itemSize: { default: null },
    gridItems: { default: void 0 },
    itemSecondarySize: { default: void 0 },
    minItemSize: { default: null },
    sizeField: { default: "size" },
    typeField: { default: "type" },
    buffer: { default: 200 },
    pageMode: { type: Boolean, default: !1 },
    prerender: { default: 0 },
    emitUpdate: { type: Boolean, default: !1 },
    disableTransform: { type: Boolean, default: !1 },
    updateInterval: { default: 0 },
    skipHover: { type: Boolean, default: !1 },
    listClass: { default: "" },
    itemClass: { default: "" }
  },
  emits: ["resize", "visible", "hidden", "update", "scroll-start", "scroll-end"],
  setup(e, { expose: n, emit: r }) {
    const t = e, i = r, c = Tt, w = Q(), S = Q(), y = Q(), g = Q(null), {
      pool: b,
      totalSize: v,
      ready: E,
      scrollToItem: C,
      scrollToPosition: x,
      getScroll: $,
      updateVisibleItems: H,
      handleScroll: D,
      handleResize: s,
      handleVisibilityChange: d
    } = $t(
      t,
      w,
      S,
      y,
      {
        onResize: () => i("resize"),
        onVisible: () => i("visible"),
        onHidden: () => i("hidden"),
        onUpdate: (z, R, h, U) => i("update", z, R, h, U)
      }
    );
    return n({
      el: w,
      scrollToItem: C,
      scrollToPosition: x,
      getScroll: $,
      updateVisibleItems: H
    }), (z, R) => rt((ee(), ue("div", {
      ref_key: "el",
      ref: w,
      class: Le(["vue-recycle-scroller", {
        ready: K(E),
        "page-mode": t.pageMode,
        [`direction-${t.direction}`]: !0
      }]),
      onScrollPassive: R[0] || (R[0] = //@ts-ignore
      (...h) => K(D) && K(D)(...h))
    }, [
      z.$slots.before ? (ee(), ue("div", {
        key: 0,
        ref_key: "before",
        ref: S,
        class: "vue-recycle-scroller__slot"
      }, [
        te(z.$slots, "before")
      ], 512)) : ge("", !0),
      (ee(), ae(Re(t.listTag), {
        style: st({ [t.direction === "vertical" ? "minHeight" : "minWidth"]: `${K(v)}px` }),
        class: Le(["vue-recycle-scroller__item-wrapper", t.listClass])
      }, {
        default: ne(() => [
          (ee(!0), ue(lt, null, ut(K(b), (h) => (ee(), ae(Rt, _e({
            key: h.nr.id,
            view: h,
            "item-tag": t.itemTag,
            style: K(E) ? [
              t.disableTransform ? { [t.direction === "vertical" ? "top" : "left"]: `${h.position}px`, willChange: "unset" } : { transform: `translate${t.direction === "vertical" ? "Y" : "X"}(${h.position}px) translate${t.direction === "vertical" ? "X" : "Y"}(${h.offset}px)` },
              {
                width: t.gridItems ? `${t.direction === "vertical" && t.itemSecondarySize || t.itemSize}px` : void 0,
                height: t.gridItems ? `${t.direction === "horizontal" && t.itemSecondarySize || t.itemSize}px` : void 0,
                visibility: h.nr.used ? "visible" : "hidden"
              }
            ] : null,
            class: ["vue-recycle-scroller__item-view", [
              t.itemClass,
              {
                hover: !t.skipHover && g.value === h.nr.key
              }
            ]]
          }, ct(t.skipHover ? {} : {
            mouseenter: () => {
              g.value = h.nr.key;
            },
            mouseleave: () => {
              g.value = null;
            }
          })), {
            default: ne((U) => [
              te(z.$slots, "default", _e({ ref_for: !0 }, U))
            ]),
            _: 2
          }, 1040, ["view", "item-tag", "style", "class"]))), 128)),
          t.items.length === 0 ? te(z.$slots, "empty", { key: 0 }) : ge("", !0)
        ]),
        _: 3
      }, 8, ["style", "class"])),
      z.$slots.after ? (ee(), ue("div", {
        key: 1,
        ref_key: "after",
        ref: y,
        class: "vue-recycle-scroller__slot"
      }, [
        te(z.$slots, "after")
      ], 512)) : ge("", !0),
      at(Mt, { onNotify: K(s) }, null, 8, ["onNotify"])
    ], 34)), [
      [K(c), K(d)]
    ]);
  }
}), Ne = /* @__PURE__ */ fe({
  inheritAttrs: !1,
  __name: "DynamicScroller",
  props: {
    items: {},
    keyField: { default: "id" },
    direction: { default: "vertical" },
    listTag: { default: "div" },
    itemTag: { default: "div" },
    minItemSize: {}
  },
  emits: ["resize", "visible"],
  setup(e, { expose: n, emit: r }) {
    const t = e, i = r, c = Q(), w = ie(() => {
      var $;
      return ($ = c.value) == null ? void 0 : $.el;
    }), {
      itemsWithSize: S,
      forceUpdate: y,
      scrollToItem: g,
      getItemSize: b,
      scrollToBottom: v,
      onScrollerResize: E,
      onScrollerVisible: C
    } = zt(
      t,
      c,
      w,
      {
        onResize: () => i("resize"),
        onVisible: () => i("visible")
      }
    );
    function x($, H, D) {
      const s = $;
      return {
        item: s.item,
        index: H,
        active: D,
        itemWithSize: s
      };
    }
    return n({
      scrollToItem: g,
      scrollToBottom: v,
      getItemSize: b,
      forceUpdate: y
    }), ($, H) => (ee(), ae($e, _e({
      ref_key: "scroller",
      ref: c,
      items: K(S),
      "min-item-size": t.minItemSize,
      direction: t.direction,
      "key-field": "id",
      "list-tag": t.listTag,
      "item-tag": t.itemTag
    }, $.$attrs, {
      onResize: K(E),
      onVisible: K(C)
    }), ft({
      default: ne(({ item: D, index: s, active: d }) => [
        te($.$slots, "default", dt(mt(x(D, s, d))))
      ]),
      empty: ne(() => [
        te($.$slots, "empty")
      ]),
      _: 2
    }, [
      $.$slots.before ? {
        name: "before",
        fn: ne(() => [
          te($.$slots, "before")
        ]),
        key: "0"
      } : void 0,
      $.$slots.after ? {
        name: "after",
        fn: ne(() => [
          te($.$slots, "after")
        ]),
        key: "1"
      } : void 0
    ]), 1040, ["items", "min-item-size", "direction", "list-tag", "item-tag", "onResize", "onVisible"]));
  }
});
function xt(e, n, r) {
  const t = he("vscrollData"), i = he("vscrollParent"), c = he("vscrollResizeObserver");
  let w = null, S = null, y = null, g = !1, b = null;
  const v = ie(() => {
    const a = u(e);
    if (t.simpleArray)
      return a.index;
    if (t.keyField in a.item)
      return a.item[t.keyField];
    throw new Error(`keyField '${t.keyField}' not found in your item. You should set a valid keyField prop on your Scroller`);
  }), E = ie(() => t.sizes[v.value] || 0), C = ie(() => u(e).active && t.active);
  function x() {
    C.value ? S !== v.value && (S = v.value, w = null, y = null, s(v.value)) : w = v.value;
  }
  function $() {
    u(e).watchData && !c ? b = L(() => u(e).item, () => {
      D();
    }, {
      deep: !0
    }) : b && (b(), b = null);
  }
  function H({ force: a }) {
    !C.value && a && (y = v.value), (w === v.value || a || !E.value) && x();
  }
  function D() {
    x();
  }
  function s(a) {
    ce(() => {
      if (v.value === a) {
        const _ = u(n);
        if (!_)
          return;
        const P = _.offsetWidth, I = _.offsetHeight;
        d(P, I);
      }
      S = null;
    });
  }
  function d(a, _) {
    const I = ~~((typeof i.direction == "object" ? i.direction.value : i.direction) === "vertical" ? _ : a);
    I && E.value !== I && z(I);
  }
  function z(a) {
    var _;
    i.$_undefinedMap[v.value] && (i.$_undefinedSizes--, i.$_undefinedMap[v.value] = void 0), t.sizes[v.value] = a, u(e).emitResize && ((_ = r == null ? void 0 : r.onResize) == null || _.call(r, v.value));
  }
  function R() {
    if (!c || g)
      return;
    const a = u(n);
    a && (c.observe(a), a.$_vs_id = v.value, a.$_vs_onResize = U, g = !0);
  }
  function h() {
    if (!c || !g)
      return;
    const a = u(n);
    a && (c.unobserve(a), a.$_vs_onResize = void 0, g = !1);
  }
  function U(a, _, P) {
    v.value === a && d(_, P);
  }
  if (L(() => u(e).watchData, () => {
    $();
  }), L(v, (a, _) => {
    const P = u(n);
    if (P && (P.$_vs_id = v.value), E.value || D(), g) {
      const I = t.sizes[_], N = t.sizes[a];
      N != null && N !== I ? z(N) : I != null && I !== N && z(I);
    }
  }), L(C, (a) => {
    E.value || (a ? i.$_undefinedMap[v.value] || (i.$_undefinedSizes++, i.$_undefinedMap[v.value] = !0) : i.$_undefinedMap[v.value] && (i.$_undefinedSizes--, i.$_undefinedMap[v.value] = !1)), c ? a ? R() : h() : a && y === v.value && x();
  }), $(), !c) {
    const a = u(e);
    if (a.sizeDependencies)
      for (const _ in a.sizeDependencies)
        L(() => u(e).sizeDependencies[_], D);
    i.$_events.on("vscroll:update", H);
  }
  return Ie(() => {
    C.value && (x(), R());
  }), Te(() => {
    i.$_events.off("vscroll:update", H), h();
  }), {
    id: v,
    size: E,
    finalActive: C,
    updateSize: x
  };
}
const We = /* @__PURE__ */ fe({
  __name: "DynamicScrollerItem",
  props: {
    item: {},
    watchData: { type: Boolean, default: !1 },
    active: { type: Boolean },
    index: { default: void 0 },
    sizeDependencies: { default: null },
    emitResize: { type: Boolean, default: !1 },
    tag: { default: "div" }
  },
  emits: ["resize"],
  setup(e, { emit: n }) {
    const r = e, t = n, i = Q();
    return xt(
      r,
      i,
      {
        onResize: (c) => t("resize", c)
      }
    ), (c, w) => (ee(), ae(Re(r.tag), {
      ref_key: "el",
      ref: i
    }, {
      default: ne(() => [
        te(c.$slots, "default")
      ]),
      _: 3
    }, 512));
  }
});
function Ct({
  idProp: e = (n) => n.item.id
} = {}) {
  const n = qe({}), r = Q(null);
  let t = null;
  const i = vt();
  if (!i)
    throw new Error("[useIdState] Must be called inside setup()");
  const c = typeof e == "function" ? () => e(i.proxy) : () => i.proxy[e];
  function w(y) {
    const g = i.proxy.$options.idState;
    if (typeof g == "function") {
      const b = g.call(i.proxy, i.proxy);
      return n[y] = b, t = y, b;
    } else
      throw new TypeError("[useIdState] Missing `idState` function on component definition.");
  }
  function S() {
    const y = c();
    y == null && console.warn(`No id found for IdState with idProp: '${e}'.`), y !== t && (n[y] || w(y), r.value = n[y]);
  }
  return L(c, (y) => {
    ce(() => {
      t = y;
    });
  }, { immediate: !0 }), S(), pt(() => {
    S();
  }), {
    idState: r
  };
}
function Dt(e, n) {
  e.component(`${n}recycle-scroller`, $e), e.component(`${n}RecycleScroller`, $e), e.component(`${n}dynamic-scroller`, Ne), e.component(`${n}DynamicScroller`, Ne), e.component(`${n}dynamic-scroller-item`, We), e.component(`${n}DynamicScrollerItem`, We);
}
const Pt = {
  version: "2.0.0-beta.10",
  install(e, n) {
    const r = { installComponents: !0, componentsPrefix: "", ...n };
    for (const t in r)
      typeof r[t] < "u" && (Ye[t] = r[t]);
    r.installComponents && Dt(e, r.componentsPrefix);
  }
};
export {
  Ne as DynamicScroller,
  We as DynamicScrollerItem,
  $e as RecycleScroller,
  Pt as default,
  zt as useDynamicScroller,
  xt as useDynamicScrollerItem,
  Ct as useIdState,
  $t as useRecycleScroller
};
